### ISSUE TITLE
Out-of-scale SR

### ISSUE THREAD

**Issue by mulkkog** [ChrisDud0257/GSASR#3]
Hello, thank you very much for sharing the code.

I tested the super-resolution with an integer scale, and it worked well. However, when I tried using a float scale, the results were not as expected.

![Image](https://github.com/user-attachments/assets/ee7812fa-e745-45e1-bb9d-d840e5e2cc73)

To handle float scale SR, I tried modifying the code slightly as follows:

if self.opt['phase'] != 'train':
    h = int(math.ceil(img_lq.shape[0] * scale))
    w = int(math.ceil(img_lq.shape[1] * scale))
    img_gt = img_gt[0:h, 0:w, :]

I used ceil to calculate the height and width, assuming this would help support non-integer scaling. Unfortunately, this approach didn’t resolve the issue.

Do you have any suggestions or comments regarding float scale SR support?
Any insights would be greatly appreciated.

Thanks again!

**Comment by mulkkog** [comment]
```yaml
name: GSASR_EDSR-Baseline_paper_bicubic_x1-4_x6.4
model_type: GSASRModel
scale: 6.4
num_gpu: 1  # set num_gpu: 0 for cpu mode
manual_seed: 0

tile_size: 48
tile_overlap: 8
tile_process: False  ### Set it to False unless the scaling factor is 2
crop_size: 4
denominator: 12

default_step_size: 1.2
cuda_rendering: True
mode: 'scale_modify'

if_dmax: True
dmax_mode: 'fix'
dmax: 0.1

datasets:
  test_1:  # the 1st test dataset
    name: Set5
    type: PairedImageDataset
    dataroot_gt: /home/jijang/ssd_data/projects/GSASR_new/datasets/1_Image_SR/test/Set5/HR
    dataroot_lq: /home/jijang/ssd_data/projects/GSASR_new/datasets/1_Image_SR/test/Set5/LR_bicubic/X6.4
    filename_tmpl: '{}x6.4'
    io_backend:
      type: disk

# network structures
network_g:
  type: EDSRNOUP
  num_in_ch: 3
  num_out_ch: 3
  num_feat: 64
  num_block: 16
  upscale: 4
  res_scale: 1

network_fea2gs:
  type: Fea2GS
  inchannel: 64
  channel: 180
  num_heads: 6
  num_crossattn_blocks: 1
  num_crossattn_layers: 2
  num_selfattn_blocks: 6
  num_selfattn_layers: 6
  num_gs_seed: 144
  gs_up_factor: 1.0
  window_size: 12
  img_range: 1.0
  shuffle_scale1: 2
  shuffle_scale2: 2

# path
path: 
  pretrain_network_g: /home/jijang/ssd_data/projects/GSASR_new/pretrained_models/edsr-baseline/encoder.pth
  strict_load_g: True
  param_key_g: params_ema

path_fea2gs: 
  pretrain_network_fea2gs: /home/jijang/ssd_data/projects/GSASR_new/pretrained_models/edsr-baseline/decoder.pth
  strict_load_fea2gs: True
  param_key_fea2gs: params_ema

# validation settings
val:
  save_img: True
  suffix: 'GSASREDSRBbicubicx6.4'  # add suffix to saved images, if None, use exp name

  metrics:
    psnr: # metric name, can be arbitrary
      type: calculate_psnr
      crop_border: 4
      test_y_channel: True
    ssim:
      type: calculate_ssim
      crop_border: 4
      test_y_channel: True
    # lpips:
    #   type: calculate_lpips
    #   crop_border: 4
    # dists:
    #   type: calculate_dists
    #   crop_border: 4
```


This is the yml file for the test I used

**Comment by mulkkog** [comment]
Problems occur even at integer scales above 4×. Could it be that the way I'm generating the LR images is incorrect?


```python
import os
import math
import random
from PIL import Image
import torch
import torchvision.transforms.functional as TF
import torchvision.transforms as transforms
from torchvision.transforms.functional import InterpolationMode

def resize_fn(img, size):
    img = transforms.Resize(size, InterpolationMode.BICUBIC)(img)
    return img

def generate_lr_images(gt_root, lr_root, scale, img_exts={'.png', '.jpg', '.jpeg'}):
    os.makedirs(lr_root, exist_ok=True)
    img_names = [f for f in os.listdir(gt_root) if os.path.splitext(f)[1].lower() in img_exts]

    for name in img_names:
        img_path = os.path.join(gt_root, name)
        img = Image.open(img_path).convert('RGB')
        img_tensor = TF.to_tensor(img)  # (C, H, W)

        h_hr = int(math.ceil(img_tensor.shape[1] / scale) * scale)
        w_hr = int(math.ceil(img_tensor.shape[2] / scale) * scale)
        img_tensor = TF.center_crop(img_tensor, (h_hr, w_hr))  # center crop instead of padding

        h_lr = int(h_hr / scale)
        w_lr = int(w_hr / scale)
        img_lr_tensor = resize_fn(img_tensor, (h_lr, w_lr))

        lr_img = TF.to_pil_image(img_lr_tensor.clamp_(0, 1))
        base, ext = os.path.splitext(name)
        out_name = f"{base}x{scale:.1f}{ext}"
        lr_img.save(os.path.join(lr_root, out_name))

        print(f"Saved: {out_name}")

gt_root = '/home/jijang/ssd_data/projects/ContinuousSR/data/1_Image_SR/test/Set5/HR'
lr_root = '/home/jijang/ssd_data/projects/ContinuousSR/data/1_Image_SR/test/Set5/LR_bicubic/X12'

scale = 12

generate_lr_images(gt_root, lr_root, scale)
```

**Comment by ChrisDud0257** [comment]
> Problems occur even at scales above 4×. Could it be that the way I'm generating the LR images is incorrect?
> 
> Problems occur even at scales above 4×. Could it be that the way I'm generating the LR images is incorrect?
> 
> import os
> import math
> import random
> from PIL import Image
> import torch
> import torchvision.transforms.functional as TF
> import torchvision.transforms as transforms
> from torchvision.transforms.functional import InterpolationMode
> 
> def resize_fn(img, size):
>     img = transforms.Resize(size, InterpolationMode.BICUBIC)(img)
>     return img
> 
> def generate_lr_images(gt_root, lr_root, scale, img_exts={'.png', '.jpg', '.jpeg'}):
>     os.makedirs(lr_root, exist_ok=True)
>     img_names = [f for f in os.listdir(gt_root) if os.path.splitext(f)[1].lower() in img_exts]
> 
>     for name in img_names:
>         img_path = os.path.join(gt_root, name)
>         img = Image.open(img_path).convert('RGB')
>         img_tensor = TF.to_tensor(img)  # (C, H, W)
> 
>         h_hr = int(math.ceil(img_tensor.shape[1] / scale) * scale)
>         w_hr = int(math.ceil(img_tensor.shape[2] / scale) * scale)
>         img_tensor = TF.center_crop(img_tensor, (h_hr, w_hr))  # center crop instead of padding
> 
>         h_lr = int(h_hr / scale)
>         w_lr = int(w_hr / scale)
>         img_lr_tensor = resize_fn(img_tensor, (h_lr, w_lr))
> 
>         lr_img = TF.to_pil_image(img_lr_tensor.clamp_(0, 1))
>         base, ext = os.path.splitext(name)
>         out_name = f"{base}x{scale:.1f}{ext}"
>         lr_img.save(os.path.join(lr_root, out_name))
> 
>         print(f"Saved: {out_name}")
> 
> gt_root = '/home/jijang/ssd_data/projects/ContinuousSR/data/1_Image_SR/test/Set5/HR'
> lr_root = '/home/jijang/ssd_data/projects/ContinuousSR/data/1_Image_SR/test/Set5/LR_bicubic/X6.4'
> 
> scale = 6.4
> 
> generate_lr_images(gt_root, lr_root, scale)

I think you should save the center-cropped 'img_tensor' as you GT image. During inference, if you still use the original img_tensor (without center-crop) as GT, then the LR image is not strictly aligned with the GT image!
